Mathlib/Topology/Algebra/Order/LiminfLimsup.lean

```diff
@@ -520,6 +520,121 @@ theorem limsup_eq_tendsto_sum_indicator_atTop (R : Type*) [StrictOrderedSemiring
+    limsup (u + v) f ≤ (limsup u f) + limsup v f := by
+  have h := isCoboundedUnder_le_add h₁ h₃
+  have h' := isBoundedUnder_le_add h₂ h₄
+  refine le_add_of_forall_lt fun a a_u b b_v ↦ ?_
+  rw [limsup_le_iff h h']
+  intro c c_ab
+  filter_upwards [eventually_lt_of_limsup_lt a_u, eventually_lt_of_limsup_lt b_v] with x a_x b_x
+  exact (add_lt_add a_x b_x).trans c_ab
+
+lemma le_liminf_add (h₁ : IsBoundedUnder (fun x1 x2 ↦ x1 ≥ x2) f u := by isBoundedDefault)
+    (h₂ : IsBoundedUnder (fun x1 x2 ↦ x1 ≤ x2) f u := by isBoundedDefault)
+    (h₃ : IsBoundedUnder (fun x1 x2 ↦ x1 ≥ x2) f v := by isBoundedDefault)
+    (h₄ : IsCoboundedUnder (fun x1 x2 ↦ x1 ≥ x2) f v := by isBoundedDefault) :
+    (liminf u f) + liminf v f ≤ liminf (u + v) f := by
+  have h := isCoboundedUnder_ge_add h₂ h₄
+  have h' := isBoundedUnder_ge_add h₁ h₃
+  refine add_le_of_forall_lt fun a a_u b b_v ↦ ?_
+  rw [le_liminf_iff h h']
+  intro c c_ab
+  filter_upwards [eventually_lt_of_lt_liminf a_u, eventually_lt_of_lt_liminf b_v] with x a_x b_x
+  exact c_ab.trans (add_lt_add a_x b_x)
+
+lemma liminf_add_le (h₁ : IsBoundedUnder (fun x1 x2 ↦ x1 ≥ x2) f u := by isBoundedDefault)
+    (h₂ : IsBoundedUnder (fun x1 x2 ↦ x1 ≤ x2) f u := by isBoundedDefault)
+    (h₃ : IsBoundedUnder (fun x1 x2 ↦ x1 ≥ x2) f v := by isBoundedDefault)
+    (h₄ : IsCoboundedUnder (fun x1 x2 ↦ x1 ≥ x2) f v := by isBoundedDefault) :
+    liminf (u + v) f ≤ (limsup u f) + liminf v f := by
+  have h := isCoboundedUnder_ge_add h₂ h₄
+  have h' := isBoundedUnder_ge_add h₁ h₃
+  exact le_add_of_forall_lt fun a a_u b b_v ↦
+    (liminf_le_iff h h').2 fun _ c_ab ↦ ((frequently_lt_of_liminf_lt h₄ b_v).and_eventually
+    (eventually_lt_of_limsup_lt a_u h₂)).mono fun _ ab_x ↦ (add_lt_add ab_x.2 ab_x.1).trans c_ab
+
+end LiminfLimsupAdd
+
+section LiminfLimsupMul
+
+open Filter Real
+
+variable {f : Filter ι} [f.NeBot] {u v : ι → ℝ}
+
+lemma le_limsup_mul (h₁ : 0 ≤ᶠ[f] u) (h₂ : IsBoundedUnder (fun x1 x2 ↦ x1 ≤ x2) f u)
+    (h₃ : 0 ≤ᶠ[f] v) (h₄ : IsBoundedUnder (fun x1 x2 ↦ x1 ≤ x2) f v) :
+    (limsup u f) * liminf v f ≤ limsup (u * v) f := by
+  have h := (isBoundedUnder_of_eventually_ge (a := 0)
+    <| (h₁.and h₃).mono fun x ⟨u_0, v_0⟩ ↦ mul_nonneg u_0 v_0).isCoboundedUnder_le
+  have h' := isBoundedUnder_le_mul_of_nonneg h₁ h₂ h₃ h₄
+  have u0 : 0 ≤ limsup u f := le_limsup_of_frequently_le h₁.frequently h₂
+  have uv : 0 ≤ limsup (u * v) f :=
+    le_limsup_of_frequently_le ((h₁.and h₃).mono fun _ ⟨hu, hv⟩ ↦ mul_nonneg hu hv).frequently h'
+  refine mul_le_of_forall_lt_of_nonneg u0 uv fun a au b bv ↦ (le_limsup_iff h h').2 fun c c_ab ↦ ?_
+  refine ((frequently_lt_of_lt_limsup
+    (isBoundedUnder_of_eventually_ge h₁).isCoboundedUnder_le au.2).and_eventually
+    ((eventually_lt_of_lt_liminf bv.2 (isBoundedUnder_of_eventually_ge h₃)).and
+    (h₁.and h₃))).mono fun x ⟨x_a, ⟨x_b, u0, _⟩⟩ ↦ ?_
+  exact c_ab.trans_le (mul_le_mul x_a.le x_b.le bv.1 u0)
```

Comment:
And here and below 😱

Reply:
This one was particularly nasty... Same issue as above.